src/ring_theory/graded_algebra/radical.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import ring_theory.graded_algebra.homogeneous_ideal
+
+/-!
+
+This file contains a proof that the radical of any homogeneous ideal is a homogeneous ideal
+
+## Main statements
+
+* `homogeneous_ideal.is_prime_iff`: a homogeneous ideal `I` is prime if and only if `I` is
+  homogeneously prime, i.e. if `x, y` are homogeneous elements such that `x * y ∈ I`, then
+  at least one of `x,y` is in `I`.
+* `homogeneous_ideal.rad`: radical of homogeneous ideal is a homogeneous ideal.
+
+## Implementation details
+
+Through out this file, the indexing type `ι` of grading is assumed to be a
+`linear_ordered_cancel_add_comm_monoid`. This might be stronger than necessary and `linearith`
+does not work on `linear_ordered_cancel_add_comm_monoid`.
+
+## Tags
+
+homogeneous, radical
+-/
+
+
+open_locale big_operators
+
+section linear_ordered_cancel_add_comm_monoid
+
+variables {ι R A : Type*}
+variables [comm_semiring R] [comm_ring A] [algebra R A]
+variables [linear_ordered_cancel_add_comm_monoid ι] [decidable_eq ι]
+variables {𝒜 : ι → submodule R A} [graded_algebra 𝒜]
+
+lemma ideal.is_homogeneous.is_prime_of_homogeneous_mem_or_mem
+  {I : ideal A} (hI : I.is_homogeneous 𝒜) (I_ne_top : I ≠ ⊤)
+  (homogeneous_mem_or_mem : ∀ {x y : A},
+    set_like.is_homogeneous 𝒜 x → set_like.is_homogeneous 𝒜 y
+    → (x * y ∈ I → x ∈ I ∨ y ∈ I)) : ideal.is_prime I :=
+⟨I_ne_top, begin
+  intros x y hxy, by_contradiction rid,
+  obtain ⟨rid₁, rid₂⟩ := not_or_distrib.mp rid,
+  letI : Π (x : A),
+    decidable_pred (λ (i : ι), graded_algebra.proj 𝒜 i x ∉ I) := λ x, classical.dec_pred _,
+  letI : Π i (x : 𝒜 i), decidable (x ≠ 0) := λ i x, classical.dec _,
+  set set₁ := (graded_algebra.support 𝒜 x).filter (λ i, graded_algebra.proj 𝒜 i x ∉ I) with set₁_eq,
+  set set₂ := (graded_algebra.support 𝒜 y).filter (λ i, graded_algebra.proj 𝒜 i y ∉ I) with set₂_eq,
+  have set₁_nonempty : set₁.nonempty,
+  { rw finset.filter_nonempty_iff,
+    contrapose! rid₁,
+    rw ← graded_algebra.sum_support_decompose 𝒜 x,
+    apply ideal.sum_mem _ rid₁,},
+  have set₂_nonempty : set₂.nonempty,
+  { rw finset.filter_nonempty_iff,
+    contrapose! rid₂,
+    rw ← graded_algebra.sum_support_decompose 𝒜 y,
+    apply ideal.sum_mem _ rid₂, },
+  set max₁ := set₁.max' set₁_nonempty with max₁_eq,
+  set max₂ := set₂.max' set₂_nonempty with max₂_eq,
+  have mem_max₁ := finset.max'_mem set₁ set₁_nonempty,
+  have mem_max₂ := finset.max'_mem set₂ set₂_nonempty,
+  replace hxy : ∀ (i : ι), (graded_algebra.decompose 𝒜 (x * y) i : A) ∈ I := λ i, hI i hxy,
+  specialize hxy (max₁ + max₂),
+  have eq :=
+    calc  graded_algebra.proj 𝒜 (max₁ + max₂) (x * y)
+        = ∑ ij in ((graded_algebra.support 𝒜 x).product (graded_algebra.support 𝒜 y)).filter
```

Comment:
I think you can improve readability by starting your proof with
```
let \pi := \lam i, graded_algebra.proj 𝒜 i,
let S := \lam x, graded_algebra.support 𝒜 x,
```

Reply:
Or even just with `open graded_algebra` so that you can use `proj 𝒜` and `support 𝒜`